From 4e52f15524547ceef459fd25907a5cff2e0f1018 Mon Sep 17 00:00:00 2001
From: gaolez <quic_gaolez@quicinc.com>
Date: Fri, 25 Nov 2022 11:38:36 +0800
Subject: [PATCH] iw: Modify frequency to channel conversion for 2G

With the addition of new 2.5M step channels in 2.4G spectrum, new
formulae are added to convert the frequency to channel in iw.

Signed-off-by: Vignesh Mohan <quic_vignmoha@quicinc.com>
Signed-off-by: Gaole Zhang  <quic_gaolez@quicinc.com>

:100644 100644 886fef2 c071eee M	util.c
---
 util.c | 38 ++++++++++++++++++++++++++++++++++----
 1 file changed, 34 insertions(+), 4 deletions(-)

diff --git a/util.c b/util.c
index 886fef2..90bbca9 100644
--- a/util.c
+++ b/util.c
@@ -190,13 +190,43 @@ int ieee80211_channel_to_frequency(int chan, enum nl80211_band band)
 	return 0; /* not supported */
 }
 
-int ieee80211_frequency_to_channel(int freq)
+#define IS_2GHZ_STEP(freq) (((freq) - 2407) % 5)
+int freq2chan2ghz(int freq)
 {
-	/* see 802.11-2007 17.3.8.3.2 and Annex J */
+	int start_chan = 0;
+	int start_freq = 2407;
+
+	/* Channel 14 is always special */
 	if (freq == 2484)
 		return 14;
-	else if (freq < 2484)
-		return (freq - 2407) / 5;
+
+	if (freq < 2412 || freq > 2492)
+		return 190 + (freq - 2399) * 2 / 5;
+
+	if (IS_2GHZ_STEP(freq)) {
+		/* For all channels which are not spaced by multiples of
+		 * 5MHz when starting from 2407.
+		 */
+		start_chan = 200;
+		start_freq = 2399;
+	} else if (freq >= 2477) {
+		/* If channel is spaced with 5MHz gap from 2407,
+		 * they are standard IEEE channels (1 - 13) till 2472.
+		 * Frequencies after 2472 will have channel numbers starting
+		 * from 221.
+		 */
+		start_chan = 221;
+		start_freq = 2477;
+	}
+
+	return start_chan + (freq - start_freq) / 5;
+}
+
+int ieee80211_frequency_to_channel(int freq)
+{
+	/* see 802.11-2007 17.3.8.3.2 and Annex J */
+	if (freq <= 2499)
+		return freq2chan2ghz(freq);
 	else if (freq >= 4910 && freq <= 4980)
 		return (freq - 4000) / 5;
 	else if (freq <= 45000) /* DMG band lower limit */
-- 
2.25.1

